cdragontoolbox/tftdata.py

```diff
@@ -65,17 +67,70 @@ def replace_in_data(data):
                     if key in data and data[key] in replacements:
                         data[key] = replacements[data[key]]
 
-            for data in template["champs"]:
+            for data in [template["sets"][x]["traits"] for x in template["sets"]]:
+                replace_in_data(data)
+            for data in [template["sets"][x]["champions"] for x in template["sets"]]:
                 replace_in_data(data)
                 if "ability" in data:
                     replace_in_data(data["ability"])
             for data in template["items"]:
                 replace_in_data(data)
-            for data in template["traits"]:
-                replace_in_data(data)
 
             with open(os.path.join(output, f"{lang}.json"), "w", encoding="utf-8") as f:
-                json.dump(template, f, cls=NaiveJsonEncoder, indent=4, sort_keys=True)
+                json.dump(template, f, cls=NaiveJsonEncoder, indent=4)
+
+    def generate_correct_set_info(self, sets, traits, champs):
+
+        set_map = {}
+        for key, value in sets.items():
+            current_set = {"name": value["name"]}
+
+            set_champs = []
+            for char_name in value["internal"]["characters"]:
+                for real_char in champs:
+                    if char_name == real_char["internal"]["name"]:
+                        set_champs.append(real_char)
+
+            set_traits = set([y for x in set_champs for y in x["traits"]])
+            set_traits = [x for x in traits for y in set_traits if y == x["internal"]["hash"]]
+
+            for champ in set_champs:
+                champ["traits"] = [x["name"] for x in traits if x["internal"]["hash"] in champ["traits"]]
+
+            current_set["traits"] = set_traits
+            current_set["champions"] = set_champs
+            set_map[key] = current_set
+
+        [x.pop("internal") for y in set_map for x in set_map[y]["champions"]]
+        [x.pop("internal") for y in set_map for x in set_map[y]["traits"]]
+
+        return set_map
+
+    def parse_lookup_table(self, map22):
+        return {x.path: x.getv("name") for x in map22.entries if x.type == "Character"}
+
+    def parse_sets(self, map22, character_lookup):
+        set_collection = [x for x in map22.entries if x.type == 0x438850FF]
+        sets = {}
+
+        for item in set_collection:
+            char_list = item.getv("characterLists")[0]
+            set_info = item[0xD2538E5A].value
+            set_no = set_info["SetNumber"]["mValue"].value
+            set_name = set_info["SetName"]["mValue"].value
+            set_characters = []
+
+            character_lists = [x for x in map22.entries if x.type == "MapCharacterList"]
```

Comment:
No need to recompute this in every iteration. It will not change.

And I think you could create a dict indexed by `x.path` directly, to speed-up looking (the `for` loop below).
Unless there are duplicate `x.path` (I don't know if it's possible).